### ISSUE TITLE
Support Python 3

### ISSUE THREAD

**Issue by jhamrick** [NYUCCL/psiTurk#215]
Python 2 is being deprecated in a few years, and more and more projects are starting to support Python 3 (some are even Python 3 only, at this point). It would be great if psiTurk migrated over as well. I think this shouldn't be too difficult--I think most of the transition can be done with `six` and by fixing print statements.

Probably, however, it would be good to increase test coverage first so we know if anything is being broken during the transition.

**Comment by suchow** [comment]
This may be hard because `boto` has support for MTurk only for Python 2, and the development team there has no plan to change that. See https://github.com/boto/boto3/issues/174.

**Comment by jhamrick** [comment]
Booo.

**Comment by fredcallaway** [comment]
I wonder if boto is even necessary. It seems to me that the boto module is just a glorified wrapper around the [mturk API](http://docs.aws.amazon.com/AWSMechTurk/latest/AWSMturkAPI/Welcome.html). There's another (albeit less glorified) [wrapper](https://github.com/ctrlcctrlv/mturk-python). I wonder how difficult it would be to switch to this wrapper.

In the short term, we could work towards Python3 compatibility by writing 3-compatible code with `__future__` imports and the six module.

**Comment by suchow** [comment]
Update: boto3 now has MTurk support, so I don't think there are any remaining blockers.

**Comment by fredcallaway** [comment]
Bump. Is there an active development effort towards Python3 support? (Python 2.7 End Of Life is 2020!)

Is there a list of items that need to be checked off? My sense is that the main task is moving from boto2 to boto3. I'm happy to take this on, but I'd need some guidance on testing procedures.

**Comment by twiecki** [comment]
bump.

**Comment by twiecki** [comment]
@fredcallaway fwiw you probably just need to update the dependency to boto3 and run `2to3` on the codebase.

**Comment by fredcallaway** [comment]
I was unable to run test_psiturk.py, and I don't want to take on 2to3 conversion without working unit tests.

**Comment by fohria** [comment]
any update on this?

**Comment by deargle** [comment]
@fredcallaway python tests are run with `pytest`.

@fohria help wanted!

**Comment by fohria** [comment]
@deargle i'm happy to help if i can, but i'm not sure where to begin. is writing more tests the first thing needed? if so i'd need some guidance as i don't have much experience with that, i'm likely what they'd call a messy "academic coder" :)

**Comment by fredcallaway** [comment]
As @twiecki suggested, I think the first thing to do would be running 2to3 (you should already have this command). Then try running the tests, seeing what breaks, and try to fix it.

**Comment by fredcallaway** [comment]
One question for @deargle: Should a python 3 effort be made as a branch of this repo or in a separate repo? Do we want to target 2/3 compatibility or just have a separate version of psiturk that only supports Python 3. If I was doing the work, I'd much prefer the latter myself.

**Comment by gureckis** [comment]
We should abandon 2.7 at this point if possible

-todd

**Comment by twiecki** [comment]
Agree with @gureckis, most projects are sunsetting python 2.7 currently.

**Comment by gureckis** [comment]
More to Fred’s point— a separate branch for 3.0 compatibility that then merges back to master once it seems tested is probably the best path forward.

-todd

**Comment by deargle** [comment]
A few months ago I started a refactor and never finished -- but the mid-refactor is merged into master because reasons. I can finish that up in the coming week or so before we start the branch and then report back.

**Comment by deargle** [comment]
Reporting back here. We don't have any tests that cover the psiturk shell. I think this was because the psiturk shell touched amt. Before I finish the refactor, and before we convert to python3, it would probably be good to get better shell / core functionality coverage. I was looking into python mock-patching the aws_services functions to always return true or something instead of making actual boto calls. I'm a bit under water with other projects at the moment though, and test suites are not something I know terribly much about at the moment. Anyone proficient with mocking classes etc?

**Comment by deargle** [comment]
_The future is now._

https://github.com/NYUCCL/psiTurk/pull/352
